src/wallet/coinselection.cpp

```diff
@@ -64,6 +64,12 @@ static const size_t TOTAL_TRIES = 100000;
 
 std::optional<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& selection_target, const CAmount& cost_of_change)
 {
+    assert(selection_target > 0);
+    assert(selection_target <= MAX_MONEY);
+
+    assert(cost_of_change > 0);
+    assert(cost_of_change <= MAX_MONEY);
```

Comment:
Weak concept ACK. I think it's a good practice to explicitly communicate assumptions, but it also feels a bit far-fetched that we'd manage to screw up our internal call stack that the asserts would actually get triggered except in unit tests.

Both `selection_target` and `cost_of_change` necessarily are always greater than 0. Yet, both these values are set by our own internal functions and initialized to non-zero values. Also, they appear in every coin selection algorithm, not just BnB. So, perhaps if we wanted to assert that they are in the proper range, the assert should be placed at the `SelectCoins` level rather than in `SelectCoinsBnB`?

> @mzumsande thanks, this should fix the test. Although, ideally the fuzz data would also include values larger than 0 but less than 1 `[0, MAX_MONEY)`. I didn't know if it was worth spending a lot of time on this PR since I haven't had any type of ack though.

from `src/consensus/.amount.h`:
```
/** Amount in satoshis (Can be negative) */
typedef int64_t CAmount;
```

CAmount is an integer satoshi count. The smallest CAmount greater than 0 is 1.

Reply:
> typedef int64_t CAmount;

Thanks, good point, it's an int type.

> perhaps if we wanted to assert that they are in the proper range, the assert should be placed at the SelectCoins level rather than in SelectCoinsBnB?

I think moving this type of check higher up the call stack could makes sense.  However there is already some precedence for this type of [assertion](https://github.com/bitcoin/bitcoin/blob/496d5c5874b05594e65fbff576da7800eb17e635/src/wallet/coinselection.cpp#L82) at this level so it didn't seem out of place.  The good thing about the assertions here is that the code is a bit more modular and stands on it's own.

Also, I noticed after I created this PR, that there is another assertion for cost_of_change [here](https://github.com/bitcoin/bitcoin/blob/496d5c5874b05594e65fbff576da7800eb17e635/src/wallet/coinselection.cpp#L393), although it seems less useful to place the assertion after the coinselection process has happened.  I mentioned this in a different PR, but this entire [method call](https://github.com/bitcoin/bitcoin/blob/496d5c5874b05594e65fbff576da7800eb17e635/src/wallet/coinselection.cpp#L165) is redundant I believe since we just endup comparing the results again [here](https://github.com/bitcoin/bitcoin/blob/496d5c5874b05594e65fbff576da7800eb17e635/src/wallet/coinselection.cpp#L166).

Reply:
You could add the same asserts to the other coin selection algorithms, they all depend on these two values. Would you mind elaborating what gave you the idea to add the asserts?

Reply:
I am looking into some optimizations, however I want to be sure I know what the bounds are to prevent possible overflows or exceptions.  There some cases already that if invalid values are passed in there's the possibility to overflow, for example if `selection_target` is `INT_MAX`, then the addition of [here](https://github.com/bitcoin/bitcoin/blob/master/src/wallet/coinselection.cpp#L95) would overflow the size of `INT_MAX`.  `selection_target` and `cost_of_change` are both signed int `int64_t` and exceeding `INT_MAX` is undefined behavior.  Likewise if `selection_target` is `INT_MIN` and `cost_of_change` is negative would again result in undefined behavior.

Reply:
Since it would be the same two values getting checked in all of the algorithms for the same reason, it would be better to add them all in a single commit.

Reply:
I looked into adding the same bounds check for the other three selection algorithms however it doesn't seem like the params they use are consistent.  They all use an equivalent of `selection_target` however call it a different name `nTargetValue` for example.  `cost_of_change` doesn't seem to be passed as an argument but instead `KnapsackSolver` uses `change_target`.  These two params [change_target](https://github.com/bitcoin/bitcoin/blob/master/src/wallet/coinselection.h#L125) and [cost_of_change](https://github.com/bitcoin/bitcoin/blob/master/src/wallet/coinselection.h#L133) appear to be different.   `ApproximateBestSubset` uses `nTotalLower` which would be a different type of boundary check.  Since each algorithm appears to have unique params, maybe it makes more sense to address each separately.  

> Since it would be the same two values getting checked in all of the algorithms for the same reason, it would be better to add them all in a single commit.

It doesn't appear to me that the same two values are checked.  What do you think about addressing each selection algorithm separately to reduce the scope of this PR?

Also, does it make sense to use 0 as a default value [here](https://github.com/bitcoin/bitcoin/blob/master/src/wallet/coinselection.h#L133) for `cost_of_change`?